File: AltCover.Cake/Options.cs
using System;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Text;

using FSDotNet = AltCover.DotNet;

namespace AltCover.Cake
{
  /// <summary>
  /// Overall test run options; override defaults as required
  /// </summary>
  public class TestOptions : FSDotNet.ICLIOptions
  {
    /// <summary>
    /// Gets whether to force delete previous runs
    /// </summary>
    public virtual bool ForceDelete => false;

    /// <summary>
    /// Gets whether to fail without computing coverage on test failure
    /// </summary>
    public virtual bool FailFast => false;

    /// <summary>
    /// Gets the type of summary to show
    /// </summary>
    public virtual string ShowSummary => String.Empty;
  }

  /// <summary>
  /// Preparation stage options; override defaults as required
  /// </summary>
  public class PrepareOptions : Abstract.IPrepareOptions
  {
    /// <summary>
    /// Gets the input directories (overridden by the MSBuild integration)
    /// </summary>
#pragma warning disable IDE0079 // Remove unnecessary suppression

    [SuppressMessage("Gendarme.Rules.BadPractice",
                     "DoNotForgetNotImplementedMethodsRule",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Design",
                     "CA1065:DoNotRaiseExceptionsInUnexpectedLocations",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Naming", "CA2204:Literals should be spelled correctly",
                  Justification = "Property name")]
    public IEnumerable<string> InputDirectories => throw new NotImplementedException("InputDirectories not used");

    /// <summary>
    /// Gets the output directories (overridden by the MSBuild integration)
    /// </summary>

    [SuppressMessage("Gendarme.Rules.BadPractice",
                     "DoNotForgetNotImplementedMethodsRule",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Design",
                     "CA1065:DoNotRaiseExceptionsInUnexpectedLocations",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Naming", "CA2204:Literals should be spelled correctly",
                  Justification = "Property name")]
    public IEnumerable<string> OutputDirectories => throw new NotImplementedException("OutputDirectories not used");

    ///<summary>
    /// Corresponds to command line option `-y, --symbolDirectory=VALUE`
    ///</summary>
    public virtual IEnumerable<string> SymbolDirectories => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `-d, --dependency=VALUE`
    ///</summary>
    public virtual IEnumerable<string> Dependencies => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `-k, --key=VALUE`
    ///</summary>
    public virtual IEnumerable<string> Keys => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `--sn, --strongNameKey=VALUE`
    ///</summary>
    public virtual string StrongNameKey => String.Empty;

    ///<summary>
    /// Gets or sets the value that corresponds to command line option `-r, --report=VALUE`
    ///</summary>
    public virtual string Report => String.Empty;

    ///<summary>
    /// Corresponds to command line option `-f, --fileFilter=VALUE`
    ///</summary>
    public virtual IEnumerable<string> FileFilter => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `-s, --assemblyFilter=VALUE`
    ///</summary>
    public virtual IEnumerable<string> AssemblyFilter => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `-e, --assemblyExcludeFilter=VALUE`
    ///</summary>
    public virtual IEnumerable<string> AssemblyExcludeFilter => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `-t, --typeFilter=VALUE`
    ///</summary>
    public virtual IEnumerable<string> TypeFilter => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `-m, --methodFilter=VALUE`
    ///</summary>
    public virtual IEnumerable<string> MethodFilter => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `-a, --attributeFilter=VALUE`
    ///</summary>
    public virtual IEnumerable<string> AttributeFilter => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `-p, --pathFilter=VALUE`
    ///</summary>
    public virtual IEnumerable<string> PathFilter => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `--attributetoplevel=VALUE`
    ///</summary>
    public virtual IEnumerable<string> AttributeTopLevel => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `--typetoplevel=VALUE`
    ///</summary>
    public virtual IEnumerable<string> TypeTopLevel => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `--methodtoplevel=VALUE`
    ///</summary>
    public virtual IEnumerable<string> MethodTopLevel => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `-c, --callContext=VALUE`
    ///</summary>
    public virtual IEnumerable<string> CallContext => Array.Empty<string>();

    ///<summary>
    /// Corresponds to command line option `--reportFormat=VALUE`
    ///</summary>
    public virtual string ReportFormat => String.Empty;

    ///<summary>
    /// Corresponds to command line option `--inplace`
    ///</summary>
    public virtual bool InPlace => false;

    ///<summary>
    /// Corresponds to command line option `--save`
    ///</summary>
    public virtual bool Save => false;

    ///<summary>
    /// Corresponds to command line option `--zipfile`
    ///</summary>
    public virtual bool ZipFile => false;

    ///<summary>
    /// Corresponds to command line option `--methodpoint`
    ///</summary>
    public virtual bool MethodPoint => false;

    ///<summary>
    /// Corresponds to command line option `--single`
    ///</summary>
    public virtual bool SingleVisit => false;

    ///<summary>
    /// Corresponds to command line option `--linecover`
    ///</summary>
    public virtual bool LineCover => false;

    ///<summary>
    /// Corresponds to command line option `--branchcover`
    ///</summary>
    public virtual bool BranchCover => false;

    /// <summary>
    /// Gets the command line (overridden by the MSBuild integration)
    /// </summary>
    [SuppressMessage("Gendarme.Rules.BadPractice",
                     "DoNotForgetNotImplementedMethodsRule",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Design",
                     "CA1065:DoNotRaiseExceptionsInUnexpectedLocations",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Naming", "CA2204:Literals should be spelled correctly",
                  Justification = "Property name")]
    public IEnumerable<string> CommandLine => throw new NotImplementedException("CommandLine not used");

    /// <summary>
    /// Gets whether to expose a program return code (overridden by the MSBuild integration)
    /// </summary>
    [SuppressMessage("Gendarme.Rules.BadPractice",
                     "DoNotForgetNotImplementedMethodsRule",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Design",
                     "CA1065:DoNotRaiseExceptionsInUnexpectedLocations",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Naming", "CA2204:Literals should be spelled correctly",
                  Justification = "Property name")]
    public bool ExposeReturnCode => throw new NotImplementedException("ExposeReturnCode not used");

    ///<summary>
    /// Corresponds to command line option `--sourcelink`
    ///</summary>
    public virtual bool SourceLink => true;

    /// <summary>
    /// Gets whether to defer coverage reporting (overridden by the MSBuild integration)
    /// </summary>
    [SuppressMessage("Gendarme.Rules.BadPractice",
                     "DoNotForgetNotImplementedMethodsRule",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Design",
                     "CA1065:DoNotRaiseExceptionsInUnexpectedLocations",
                     Justification = "Means 'do not use'")]
    public bool Defer => throw new NotImplementedException("Defer not used");

    ///<summary>
    /// Corresponds to command line option `-l, --localSource`
    ///</summary>
    public virtual bool LocalSource => true;

    ///<summary>
    /// Corresponds to command line option `-v, --visibleBranches`
    ///</summary>
    public virtual bool VisibleBranches => false;

    ///<summary>
    /// Corresponds to command line option `--showstatic[=VALUE]`
    ///</summary>
    public virtual string ShowStatic => String.Empty;

    ///<summary>
    /// Corresponds to command line option `--showGenerated`
    ///</summary>
    public virtual bool ShowGenerated => false;

    ///<summary>
    /// Corresponds to command line option `-q`
    ///</summary>
    public virtual System.Diagnostics.TraceLevel Verbosity => System.Diagnostics.TraceLevel.Info;

    ///<summary>
    /// Corresponds to command line option `--trivia`
    ///</summary>
    public virtual bool Trivia => false;

    ///<summary>
    /// Corresponds to dotnet test option `/p:AltCoverOutputRoot`
    ///</summary>
    public virtual string OutputRoot => String.Empty;
  }

  /// <summary>
  /// Collection stage options; override defaults as required
  /// </summary>
  public class CollectOptions : Abstract.ICollectOptions
  {
    /// <summary>
    /// Gets the recorder directory (overridden by the MSBuild integration)
    /// </summary>
    [SuppressMessage("Gendarme.Rules.BadPractice",
                     "DoNotForgetNotImplementedMethodsRule",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Design",
                     "CA1065:DoNotRaiseExceptionsInUnexpectedLocations",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Naming", "CA2204:Literals should be spelled correctly",
                  Justification = "Property name")]
    public string RecorderDirectory => throw new NotImplementedException("RecorderDirectory not used");

    /// <summary>
    /// Gets the working directory (overridden by the MSBuild integration)
    /// </summary>
    [SuppressMessage("Gendarme.Rules.BadPractice",
                     "DoNotForgetNotImplementedMethodsRule",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Design",
                     "CA1065:DoNotRaiseExceptionsInUnexpectedLocations",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Naming", "CA2204:Literals should be spelled correctly",
                  Justification = "Property name")]
    public string WorkingDirectory => throw new NotImplementedException("WorkingDirectory not used");

    /// <summary>
    /// Gets the programe to run (overridden by the MSBuild integration)
    /// </summary>
    [SuppressMessage("Gendarme.Rules.BadPractice",
                     "DoNotForgetNotImplementedMethodsRule",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Design",
                     "CA1065:DoNotRaiseExceptionsInUnexpectedLocations",
                     Justification = "Means 'do not use'")]
    public string Executable => throw new NotImplementedException("Executable not used");

    ///<summary>
    /// Corresponds to command line option `-l, --lcovReport=VALUE`
    ///</summary>
    [SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly",
                      Justification = "Lcov is a name")]
    public virtual string LcovReport => String.Empty;

    ///<summary>
    /// Corresponds to command line option `-t, --threshold=VALUE`
    ///</summary>
    public virtual string Threshold => String.Empty;

    ///<summary>
    /// Corresponds to command line option `-c, --cobertura=VALUE`
    ///</summary>
    [SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly",
                      Justification = "Lcov is a name")]
    public virtual string Cobertura => String.Empty;

    ///<summary>
    /// Corresponds to command line option `-o, --outputFile=VALUE`
    ///</summary>
    public virtual string OutputFile => String.Empty;

    /// <summary>
    /// Gets the command line (overridden by the MSBuild integration)
    /// </summary>
    [SuppressMessage("Gendarme.Rules.BadPractice",
                     "DoNotForgetNotImplementedMethodsRule",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Design",
                     "CA1065:DoNotRaiseExceptionsInUnexpectedLocations",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Naming", "CA2204:Literals should be spelled correctly",
                  Justification = "Property name")]
    public IEnumerable<string> CommandLine => throw new NotImplementedException("CommandLine not used");

    /// <summary>
    /// Gets whether to expose a program return code (overridden by the MSBuild integration)
    /// </summary>
    [SuppressMessage("Gendarme.Rules.BadPractice",
                     "DoNotForgetNotImplementedMethodsRule",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Design",
                     "CA1065:DoNotRaiseExceptionsInUnexpectedLocations",
                     Justification = "Means 'do not use'")]
    [SuppressMessage("Microsoft.Naming", "CA2204:Literals should be spelled correctly",
                  Justification = "Property name")]
    public bool ExposeReturnCode => throw new NotImplementedException("ExposeReturnCode not used");

    ///<summary>
    /// Corresponds to command line option `--teamcity[=VALUE]`
    ///</summary>
    public virtual string SummaryFormat => String.Empty;

    ///<summary>
    /// Corresponds to command line option `-q`
    ///</summary>
    public virtual System.Diagnostics.TraceLevel Verbosity => System.Diagnostics.TraceLevel.Info;
  }
}

File: AltCover.DotNet/Options.fs
#if RUNNER
namespace AltCover
#else
namespace AltCoverFake.DotNet.Testing
#endif

open System
open System.Diagnostics.CodeAnalysis

[<RequireQualifiedAccess>]
module Options =
  [<AutoSerializable(false)>]
  [<SuppressMessage("Gendarme.Rules.Maintainability",
                    "AvoidLackOfCohesionOfMethodsRule",
                    Justification = "It's a property bag")>]
  type CLI() =
    member val ForceDelete = false with get, set
    member val FailFast = false with get, set
    member val ShowSummary = String.Empty with get, set

    interface DotNet.ICLIOptions with
      member self.ForceDelete = self.ForceDelete
      member self.FailFast = self.FailFast
      member self.ShowSummary = self.ShowSummary

  [<AutoSerializable(false)>]
  [<SuppressMessage("Gendarme.Rules.Maintainability",
                    "AvoidLackOfCohesionOfMethodsRule",
                    Justification = "It's a property bag")>]
  type Collect() =
    member val RecorderDirectory = String.Empty with get, set
    member val WorkingDirectory = String.Empty with get, set
    member val Executable = String.Empty with get, set

    [<SuppressMessage("Microsoft.Naming",
                      "CA1704:IdentifiersShouldBeSpelledCorrectly",
                      Justification = "Lcov is a name")>]
    member val LcovReport = String.Empty with get, set

    member val Threshold = String.Empty with get, set

    [<SuppressMessage("Microsoft.Naming",
                      "CA1704:IdentifiersShouldBeSpelledCorrectly",
                      Justification = "Cobertura is a name")>]
    member val Cobertura = String.Empty with get, set

    member val Packages = Seq.empty<String> with get, set

    member val OutputFile = String.Empty with get, set
    member val CommandLine = Seq.empty<String> with get, set
    member val ExposeReturnCode = true with get, set
    member val SummaryFormat = String.Empty with get, set
    member val Verbosity = System.Diagnostics.TraceLevel.Info with get, set

    interface Abstract.ICollectOptions with
      member self.RecorderDirectory =
        self.RecorderDirectory

      member self.WorkingDirectory =
        self.WorkingDirectory

      member self.Executable = self.Executable

      [<SuppressMessage("Microsoft.Naming",
                        "CA1704:IdentifiersShouldBeSpelledCorrectly",
                        Justification = "Lcov is a name")>]
      member self.LcovReport = self.LcovReport

      member self.Threshold = self.Threshold

      [<SuppressMessage("Microsoft.Naming",
                        "CA1704:IdentifiersShouldBeSpelledCorrectly",
                        Justification = "Cobertura is a name")>]
      member self.Cobertura = self.Cobertura

      member self.Packages = self.Packages

      member self.OutputFile = self.OutputFile
      member self.CommandLine = self.CommandLine

      member self.ExposeReturnCode =
        self.ExposeReturnCode

      member self.SummaryFormat = self.SummaryFormat
      member self.Verbosity = self.Verbosity

  [<AutoSerializable(false)>]
  [<SuppressMessage("Gendarme.Rules.Maintainability",
                    "AvoidLackOfCohesionOfMethodsRule",
                    Justification = "It's a property bag")>]
  [<SuppressMessage("Gendarme.Rules.Smells",
                    "AvoidLargeClassesRule",
                    Justification = "It's a property bag")>]
  type Prepare() =
    member val InputDirectories = Seq.empty<String> with get, set
    member val OutputDirectories = Seq.empty<String> with get, set
    member val SymbolDirectories = Seq.empty<String> with get, set
    member val Dependencies = Seq.empty<String> with get, set
    member val Keys = Seq.empty<String> with get, set
    member val StrongNameKey = String.Empty with get, set
    member val Report = String.Empty with get, set
    member val FileFilter = Seq.empty<String> with get, set
    member val AssemblyFilter = Seq.empty<String> with get, set
    member val AssemblyExcludeFilter = Seq.empty<String> with get, set
    member val TypeFilter = Seq.empty<String> with get, set
    member val MethodFilter = Seq.empty<String> with get, set
    member val AttributeFilter = Seq.empty<String> with get, set
    member val PathFilter = Seq.empty<String> with get, set
    member val AttributeTopLevel = Seq.empty<String> with get, set
    member val TypeTopLevel = Seq.empty<String> with get, set
    member val MethodTopLevel = Seq.empty<String> with get, set
    member val CallContext = Seq.empty<String> with get, set
    member val ReportFormat = String.Empty with get, set
    member val InPlace = false with get, set
    member val Save = true with get, set
    member val ZipFile = false with get, set
    member val MethodPoint = false with get, set
    member val SingleVisit = false with get, set
    member val LineCover = false with get, set
    member val BranchCover = false with get, set
    member val CommandLine = Seq.empty<String> with get, set
    member val ExposeReturnCode = true with get, set
    member val SourceLink = false with get, set
    member val Defer = false with get, set
    member val LocalSource = false with get, set
    member val VisibleBranches = false with get, set
    member val ShowStatic = "-" with get, set
    member val ShowGenerated = false with get, set
    member val Verbosity = System.Diagnostics.TraceLevel.Info with get, set
    member val Trivia = false with get, set
    member val OutputRoot = String.Empty with get, set
    member val Portable = false with get, set

    interface Abstract.IPrepareOptions with
      member self.InputDirectories =
        self.InputDirectories

      member self.OutputDirectories =
        self.OutputDirectories

      member self.SymbolDirectories =
        self.SymbolDirectories

      member self.Dependencies = self.Dependencies
      member self.Keys = self.Keys
      member self.StrongNameKey = self.StrongNameKey
      member self.Report = self.Report
      member self.FileFilter = self.FileFilter

      member self.AssemblyFilter =
        self.AssemblyFilter

      member self.AssemblyExcludeFilter =
        self.AssemblyExcludeFilter

      member self.TypeFilter = self.TypeFilter
      member self.MethodFilter = self.MethodFilter

      member self.AttributeFilter =
        self.AttributeFilter

      member self.PathFilter = self.PathFilter

      member self.AttributeTopLevel =
        self.AttributeTopLevel

      member self.TypeTopLevel = self.TypeTopLevel

      member self.MethodTopLevel =
        self.MethodTopLevel

      member self.CallContext = self.CallContext
      member self.ReportFormat = self.ReportFormat
      member self.InPlace = self.InPlace
      member self.Save = self.Save
      member self.ZipFile = self.ZipFile
      member self.MethodPoint = self.MethodPoint
      member self.SingleVisit = self.SingleVisit
      member self.LineCover = self.LineCover
      member self.BranchCover = self.BranchCover
      member self.CommandLine = self.CommandLine

      member self.ExposeReturnCode =
        self.ExposeReturnCode

      member self.SourceLink = self.SourceLink
      member self.Defer = self.Defer
      member self.LocalSource = self.LocalSource

      member self.VisibleBranches =
        self.VisibleBranches

      member self.ShowStatic = self.ShowStatic
      member self.ShowGenerated = self.ShowGenerated
      member self.Verbosity = self.Verbosity
      member self.Trivia = self.Trivia
      member self.OutputRoot = self.OutputRoot
      member self.Portable = self.Portable

#if RUNNER
  [<AutoSerializable(false)>]
  [<SuppressMessage("Gendarme.Rules.Maintainability",
                    "AvoidLackOfCohesionOfMethodsRule",
                    Justification = "It's a property bag")>]
  type Logging() =
    let wl (w: System.IO.TextWriter) (x: string) = w.WriteLine x
    member val Info = Action<string>(wl Console.Out) with get, set
    member val Warn = Action<string>(wl Console.Out) with get, set
    member val Failure = Action<string>(wl Console.Error) with get, set
    member val Echo = Action<string>(wl Console.Out) with get, set
    member val Verbose = Action<string>(wl Console.Out) with get, set

    interface Abstract.ILoggingOptions with
      member self.Info = self.Info
      member self.Warn = self.Warn
      member self.Failure = self.Failure
      member self.Echo = self.Echo
      member self.Verbose = self.Verbose
#endif